js/navbar/tabColor.js

```diff
@@ -243,7 +200,61 @@ function setColor (bg, fg) {
   }
 }
 
-// theme changes can affect the tab colors
-window.addEventListener('themechange', function (e) {
-  updateColorPalette()
-})
+const tabColor = {
+  initialize: function () {
+    webviews.bindEvent('page-favicon-updated', function (e, favicons) {
+      var id = webviews.getTabFromContents(this)
+      tabColor.updateFromImage(favicons, id)
+    })
+
+    // theme changes can affect the tab colors
+    window.addEventListener('themechange', function (e) {
+      tabColor.refresh()
+    })
+  },
+  updateFromImage: function (favicons, tabId) {
+    // private tabs always use a special color, we don't need to get the icon
+    if (tabs.get(tabId).private === true) {
+      return
+    }
+
+    requestIdleCallback(function () {
+      colorExtractorImage.onload = function (e) {
+        let backgroundColor = getColorFromImage(colorExtractorImage)
+        let textColor = getTextColor(backgroundColor)
+
+        let backgroundString = getRGBString(backgroundColor)
+
+        tabs.update(tabId, {
+          backgroundColor: backgroundString,
+          foregroundColor: textColor
+        })
+
+        if (tabId === tabs.getSelected()) {
+          tabColor.refresh()
+        }
+      }
+      colorExtractorImage.src = favicons[0]
+    }, {
+      timeout: 1000
+    })
+  },
+  refresh: function () {
+    var tab = tabs.get(tabs.getSelected())
+
+    if (tab.private) {
+      // private tabs have their own color scheme
+      return setColor(defaultColors.private[0], defaultColors.private[1])
+    // use the colors extracted from the page icon
+    } else if (tab.backgroundColor || tab.foregroundColor) {
```

Comment:
You don't need the `else` blocks, because you're doing `return X` in every if-block. Removing them will also flatten the function, making it flowing more straightforwardly.